FIXES wp-rocket#7556: Adds a filter to show or not the imagify banner

Miraeld opened the pull request:
# Description

Fixes wp-media/wp-rocket#7556

Add filter to control Imagify banner on Media gallery components
- Add 'wpmedia_plugin_family_show_imagify_banner' filter to allow developers to control banner visibility
- Banner is enabled by default and can be disabled by setting filter to false
- Bump version to 1.0.7

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

## Detailed scenario

### What was tested

### How to test

1. Install WP Rocket
2. Ensure Imagify is not installed or activated
3. Navigate to Media > Library or edit a post
4. Verify the Imagify promotion banner appears or not

### How to test

Use the filter `add_filter( 'wpmedia_plugin_family_show_imagify_banner', '__return_false' );` to enable/disable the banner

### Affected Features & Quality Assurance Scope

- **Imagify promotion banner**: The banner display logic now includes a filter hook, allowing third-party customization
- **Media library and post editor screens**: These are the primary screens where the banner appears
- **Default behavior**: No changes to existing functionality - banner still shows by default

## Technical description

### Documentation

This pull request introduces a new dependency for typed filter application, updates the plugin version, and enhances the logic for displaying the Imagify banner by making it filterable and type-safe. The most important changes are grouped below:

Dependency management:

* Added a new required dependency on `wp-media/apply-filters-typed` in `composer.json` to support type-safe filter application.

Plugin versioning:

* Updated the plugin version from `1.0.6` to `1.0.7` in `PluginFamily.php`.

Admin banner display logic:

* Refactored the logic for determining when to show the Imagify banner in the admin area, and introduced a new filter `wpmedia_plugin_family_show_imagify_banner` that uses the new type-safe filter application (`wpm_apply_filters_typed`). This allows developers to reliably modify the banner display behavior.
### New dependencies

`"wp-media/apply-filters-typed": "^1.0"` in `composer.json`
### Risks

None

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

Miraeld committed:
Add filter to control Imagify banner on Media gallery components

- Add 'wpmedia_plugin_family_show_imagify_banner' filter to allow developers to control banner visibility
- Banner is enabled by default and can be disabled by setting filter to false
- Bump version to 1.0.7

Fixes wp-media/wp-rocket#7556

jeawhanlee commented on src/Controller/PluginFamily.php:
```diff
@@ -467,10 +467,19 @@ static function ( $p ) {
-			return in_array( get_current_screen()->id, $allowed_screen_ids, true );
+			$can_enqueue = in_array( get_current_screen()->id, $allowed_screen_ids, true );
+		} else {
+			$can_enqueue = in_array( $page, $allowed_pages, true );
```
It would be great to use optional override pattern:
```PHP
$variable = 'default'

if ( 'met' === condition ) {
  $variable = 'new value'
}
```

Miraeld committed:
Address code review feedback

- Fix composer.json formatting (remove trailing whitespace and blank line)
- Refactor can_enqueue_admin_assets to use optional override pattern
- Revert test changes per review discussion

Miraeld commented:
I've addressed the review comments from @jeawhanlee and Copilot:

### Changes in commit 49140f8:

1. **Fixed composer.json formatting** ✅
   - Removed trailing whitespace on closing brace
   - Removed extra blank line in the `require` block

2. **Refactored `can_enqueue_admin_assets` to use optional override pattern** ✅
   - Changed from if-else to set default value first, then override if condition is met
   - This follows the recommended pattern from @jeawhanlee


The code now follows the requested patterns and formatting standards.

Miraeld committed:
Fix typo

jeawhanlee commented on src/Controller/PluginFamily.php:
```diff
@@ -12,7 +12,7 @@ class PluginFamily implements PluginFamilyInterface {
 	 * @var string
 	 */
-	private $version = '1.0.6';
+	private $version = '1.0.7';
```
```suggestion
	private $version = '1.0.8';
```
We released 1.0.7 previously without changing this version number.

Commit:
Updated version to 1.0.8

Miraeld committed:
Apply filter to block editor assets

- Add wpmedia_plugin_family_show_imagify_banner filter to enqueue_assets method
- This ensures the filter works for both classic media library and block editor media modal
- Fixes issue where banner was not hidden in block editor when filter returned false

Miraeld committed:
Change version in @since

Miraeld committed:
Fix returned type filter

jeawhanlee commented on src/Controller/PluginFamily.php:
The filter is duplicated across 2 methods in the same class, it would make sense to harmonize them into a single method that can be shared wherever we need to filter values in the class. WDYT?

Miraeld committed:
Refactor filter logic into shared method

- Create should_show_imagify_banner() method to centralize filter logic
- Remove duplicate filter code from enqueue_assets() and can_enqueue_admin_assets()
- Both methods now use the shared method for consistency and maintainability

Miraeld commented:
@jeawhanlee Good point! 

I've refactored the code to eliminate the duplication. Changes in commit [d935f98](https://github.com/wp-media/plugin-family/commit/d935f98):

**Created a shared method:**
```php
private function should_show_imagify_banner( bool $default_value = true ): bool
```

This method:
- Centralizes the filter logic in one place
- Accepts a default value parameter (useful for `can_enqueue_admin_assets` which passes the calculated `$can_enqueue` value)
- Returns the filtered result

**Both methods now use it:**
- `enqueue_assets()` → calls `should_show_imagify_banner()` with default `true`
- `can_enqueue_admin_assets()` → calls `should_show_imagify_banner($can_enqueue)` with the calculated value

Mai-Saad commented:
Working as expected. existing issue(s) already on trunk. https://github.com/wp-media/wp-rocket/issues/7962
